### ISSUE TITLE
NativeMethods.loss_metric_operator_matrixs fails with System.AccessViolationException on x86 platform

### ISSUE THREAD

**Issue by Megadino** [takuya-takeuchi/DlibDotNet#136]
# Summary of your issue
When using DlibDotNet.Dnn.LossMetric class and call Operator() method, then NativeMethods.loss_metric_operator_matrixs fails with System.AccessViolationException on x86 platform, while it works perfect with x64 version of DlibDotNet

# Environment
Windows 7 x64 Service Pack 1
Visual Studio 2017
DlibDotNet 19.17.0.20190813, binaries from folder DlibDotNet.19.17.0.20190813\runtimes\win-x86\native

## Example code:
take FaceDetection WinForms sample project, set x86 target platform, add next lines:

```
var predictor = ShapePredictor.Deserialize("shape_predictor_5_face_landmarks.dat");
                    var dnn = DlibDotNet.Dnn.LossMetric.Deserialize("dlib_face_recognition_resnet_model_v1.dat");

                    var chips = new List<Matrix<RgbPixel>>();
                    var faces = new List<Rectangle>();
                    foreach (var face in dets)
                    {
                        var shape = predictor.Detect(image, face);
                        var faceChipDetail = Dlib.GetFaceChipDetails(shape, 150, 0.25);
                        var faceChip = Dlib.ExtractImageChip<RgbPixel>(image, faceChipDetail);
                        var matrix = new Matrix<RgbPixel>(faceChip);
                        chips.Add(matrix);
                        faces.Add(face);                        
                    }
                    var descriptors = dnn.Operator(chips);
```
right after 

```
                    this.pictureBox.Invoke(new Action(() =>
                    {
                        this.pictureBox.Image?.Dispose();
                        this.pictureBox.Image = result;
                    }));

```
then it fails on the line

`var descriptors = dnn.Operator(chips);`

with **System.AccessViolationException**

and stack trace is:
```
DlibDotNet.NativeMethods.loss_metric_operator_matrixs(IntPtr obj, Int32 type, MatrixElementType element_type, IntPtr matrixs, Int32 templateRows, Int32 templateColumns, UInt64 batchSize, IntPtr& ret)
   в DlibDotNet.Dnn.LossMetric.Operator[T](IEnumerable`1 images, UInt64 batchSize)
   в FaceDetection.MainForm.BackgroundWorkerOnDoWork(Object sender, DoWorkEventArgs doWorkEventArgs) в E:\DlibDotNet-master\DlibDotNet-master\examples\WinForms\FaceDetection\MainForm.cs

```

**Comment by zlyxm** [comment]
Me too

```
          var img = FaceRecognition.LoadImageFile(textBox1.Text,Mode.Greyscale);
            var land = fr.FaceLocations(img);
            var ens = fr.FaceEncodings(img, land);
            foreach (var v in ens)
//System.AccessViolationException: Attempted to read or write protected memory This is often an indication that other memory is corrupt
            {
                 richTextBox1.AppendText(v.Size.ToString() + Environment.NewLine);
                //richTextBox1.AppendText(v.Count.ToString() + Environment.NewLine);
            }
```
I checked the definition,
Is triggered by NativeMethods.loss_metric_operator_matrixs

 I am sorry that I can't help you

**Comment by takuya-takeuchi** [comment]
@Megadino and @zlyxm 
I was able to reproduce issue.
But let me have a time to check.
DlibDotNet.Dnn is huge library and it takes many build time as Debug build.
And this library uses many macro. It means that it is difficult to use step in debug.

So I have to delete unnecessary code and disassemble macro to start debug.

**Comment by takuya-takeuchi** [comment]
# Note

![image](https://user-images.githubusercontent.com/6241854/65812837-2c28d500-e208-11e9-8bfa-de23a0112791.png)

**out IntPtr** is treated as 0x000000 in C++ side. It is very wired.
But it works fine x64.

**Comment by takuya-takeuchi** [comment]
OK. I understand why out value is nullptr.

**Comment by takuya-takeuchi** [comment]
Oh, other issue was same cause.

**Comment by zlyxm** [comment]
Fabulous，I am happy to support you, though not much.
 but I didn't find the entrance. Maybe I can visit 名古屋!

**Comment by takuya-takeuchi** [comment]
@Megadino and @zlyxm 
I released 19.18.0.20190928
https://www.nuget.org/packages/DlibDotNet/19.18.0.20190928

It should work fine.

**Comment by Megadino** [comment]
@takuya-takeuchi thank you so much for your work!
